pallets/msa/src/lib.rs

```diff
@@ -1966,6 +2546,201 @@ impl<T: Config> Pallet<T> {
+			if !Self::is_valid_bcp47(code_str) {
+				return Err(Error::<T>::InvalidBCP47LanguageCode.into());
+			}
+		}
+
+		Ok(())
+	}
+
+	/// Checks if the provided BCP-47 language code is valid.
+	fn is_valid_bcp47(code: &str) -> bool {
+		// Must not be empty
+		if code.is_empty() {
+			return false;
+		}
+		// No leading, trailing, or consecutive dashes
+		if code.starts_with('-') || code.ends_with('-') || code.contains("--") {
+			return false;
+		}
+		for part in code.split('-') {
+			let len = part.len();
+			if len < 2 || !part.chars().all(|c| c.is_ascii_alphanumeric()) {
+				return false;
+			}
+		}
+		true
+	}
+
+	/// Retrieves the provider or application context including logos and localized name if any
+	pub fn get_provider_application_context(
+		provider_id: ProviderId,
+		application_id: Option<ApplicationIndex>,
+		locale: Option<Vec<u8>>,
+	) -> Option<ProviderApplicationContext> {
+		let bounded_locale = locale.and_then(|loc| BoundedVec::try_from(loc).ok());
+		let provider_or_application_registry = match application_id {
+			Some(app_id) => ProviderToApplicationRegistry::<T>::get(provider_id, app_id)?,
+			None => ProviderToRegistryEntryV2::<T>::get(provider_id)?,
+		};
+		let default_name = provider_or_application_registry.default_name.to_vec();
+		let default_logo_cid = provider_or_application_registry.default_logo_250_100_png_cid;
+		// Default logo bytes
+		let default_logo_250_100_png_bytes: Option<Vec<u8>> =
+			ApprovedLogos::<T>::get(default_logo_cid).map(|bv| bv.to_vec());
+		let mut localized_name: Option<Vec<u8>> = None;
+		// Localized logo bytes if any
+		let localized_logo_250_100_png_bytes: Option<Vec<u8>> = bounded_locale.and_then(|locale| {
+			// Localized name if any
+			localized_name = provider_or_application_registry
+				.localized_names
+				.get(&locale)
+				.map(|bv| bv.to_vec());
+
+			provider_or_application_registry
+				.localized_logo_250_100_png_cids
+				.get(&locale)
+				.and_then(|cid| ApprovedLogos::<T>::get(cid).map(|bv| bv.to_vec()))
+		});
+
+		Some(ProviderApplicationContext {
+			default_name,
+			provider_id,
+			application_id,
+			default_logo_250_100_png_bytes,
+			localized_logo_250_100_png_bytes,
+			localized_name,
+		})
+	}
+
+	/// Refund weights for logos not removed from worst case scenario
+	fn refund_logo_removal_weight_by_count(
+		total_logos_removed: u32,
+		base_weight: Weight,
+	) -> DispatchResultWithPostInfo {
+		// Weight adjustment: refund weight for logos that were NOT removed
+		let max_logos_benchmark_assumed = T::MaxLocaleCount::get();
+		let removal_over_charged_by =
+			max_logos_benchmark_assumed.saturating_sub(total_logos_removed);
+		let weight_per_logo_removal = T::DbWeight::get().writes(1);
+		let weight_to_refund =
+			weight_per_logo_removal.saturating_mul(removal_over_charged_by as u64);
+		let actual_weight_used = base_weight.saturating_sub(weight_to_refund);
+		Ok(PostDispatchInfo { actual_weight: Some(actual_weight_used), pays_fee: Pays::Yes })
+	}
```

Comment:
@aramikm  so in  both `update_provider/application` benchmarks I am not computing weight for worse case and refunding the weight that was not used (if provider upserts less than `MaxLocaleCount`), I have kept one paramters in benchmarks `n` which is total localized names as it is stored in registry and not separately